castervoice/asynch/corrections/manager.py

```diff
@@ -0,0 +1,103 @@
+"""
+Manager for the corrections window. This module runs in the 
+Caster process. Its primary purpose is to initialize, manage, and communicate
+with the process that is actively displaying the window.
+
```

Comment:
Why not use the xmlrpc client/server mechanism that other caster asynch subprocesses use? This would make it easier to integrate remedy dialogue in a consolidated GUI process eventually, as opposed to having each dialogue/window running a separate GUI application.

Reply:
I originally implemented it that way, but it was way too slow. XMLRPC + a TCP Server is much much slower than Python pickle v5 + a system pipe. Actually, I think I'm still on pickle v4 in this PR, because v5 has to be explicitly enabled, but v4 is super fast, too. v5 adds some interesting shared memory features, but I haven't researched how it works.

I had a similar idea to port the other GUI windows to a single application, but I recommend putting them behind multiprocessing. Under the hood, multiprocessing does the same thing as the other windows (spins up a separate Python process, at least on Windows; on Unix systems it uses a fork, which is much faster), but it manages all the weird edge cases for you. 

As described in the PR comment, another thing I did in this PR was to load the Window once, and keep it hidden in the background instead of constructing it from scratch on every command. This gave a modest speed boost, but for something like Rainbow Grid, it would be a huge improvement. On my 4K screens, Ranbow takes seconds to load and is therefore unusable. If the window was preloaded, it would snap up instantly.

Reply:
> I originally implemented it that way, but it was way too slow. XMLRPC + a TCP Server is much much slower than Python pickle v5 + a system pipe. Actually, I think I'm still on pickle v4 in this PR, because v5 has to be explicitly enabled, but v4 is super fast, too. v5 adds some interesting shared memory features, but I haven't researched how it works.

Are other caster asynch apps also slow on your end? It is surprising to me, as I never observed noticeable delays by xmlrpc.

Reply:
The various mouse navigation windows are unusable for me, although most of that is from rendering the windows, which takes between 1 and 5 seconds depending on which grid is being used; rainbow is longest.

In my profiling for the remedy window (I didn't keep numbers, sorry) the time delays were in this order:

 *  Time it takes to start up the interpreter
 *  Time it takes to render and display the window
 *  Time it takes for Dragon to recognize the command
 *  Time it takes to send commands

I optimized each in turn as I've described, other than the Dragon recognition which is outside my control (switching from Normal mode to Command mode helps, but that interfered with my dictation as I kept having to remember whether to insert an "ace" after each "say", "cop" or "slip"). The biggest win on sending commands was reducing the size and number of commands (calculating the choices in the subprocess instead of passing them across the pipe, for example) more than changing the transport, but since I'd already done that part I left it in. If you leave the other optimizations in place, switching it back to XMLRpc should only be a few minutes work as they are both just bidirectional pipes.

Something like the corrections window needs to be instantly responsive or else it interrupts concentration too much to be useful. It's still a little too slow for my taste, especially when it comes to hiding the window after selecting the choice. I *think* I've got it to the point where the highest percentage of the time is consumed by Dragon deciding whether it recognized a command, but it may be worth doing some finer grained profiling and compiling certain pieces of it with Cython.

The main drawback of multiprocessing is that it explicitly depends on Python. Your fastest option would be to write the GUI half of the applications in a low-level language like C++ or Rust and then use an IPC pipe to communicate between the two using a compact protocol such as CBOR. But I don't think that will give enough of an improvement to justify the maintenance overhead, and the time would probably be better spent profiling and optimizing natlink or finding some way to speed Dragon itself up.

Overall, I've been pretty disappointed in the corrections window experiment. It does a pretty good job of offering useful corrections, but even fully optimized, it interrupts my flow enough that I lose a minute of concentration for every mistake Dragon makes. This seems to be an unfixable function of the way my brain works, and I've been recently investing my time in ways to get my typing back instead. Spending a few hours coding the corrections window would probably have been much better spent doing a lot of hand and forearm yoga!

Reply:
Okay why not try it this way. As an option have the Remedy dialogue box stay in the forefront but not in focus. That way can sit in the corner of the screen and update as you work with dictation. That would that would reduce `render and display the window`.

Reply:
That'd be brilliant; something like how the Serenade window works. I'm imagining that it normally shows just the alternate interpretations the underlying speech engine gives, perhaps with some different capitalization options. But if the user says "fix <phrase>" it would drop into something like this version.

I had poked a bit at the alternate phrasing diff on the dragonfly grammar, but I couldn't figure out how to patch it through. Definitely worth it, though.

Reply:
> I couldn't figure out how to patch it through

Patching it through in terms of updating a grammar during runtime or updating the GUI?

Reply:
of where to access the grammar value from within a rule.

Reply:
Okay I think I know what you're saying. You have a variable that's defined in a grammar. It's only evaluated once during rule initialization. Therefore you can't access the updated value in the grammar because it has not been evaluated.

So rather than doing the standard `Function(myfunction)`use a lambda.
`Function(lambda n: say_number(n))`